pkg/kubectl/cmd/util/factory.go

```diff
@@ -230,12 +230,53 @@ func makeInterfacesFor(versionList []unversioned.GroupVersion) func(version unve
 	}
 }
 
+func NewShortcutExpander(clients *ClientCache, delegate meta.RESTMapper) kubectl.ShortcutExpander {
+	defaultMapper := kubectl.ShortcutExpander{All: nil, RESTMapper: delegate}
+	if clients == nil {
+		return defaultMapper
+	}
+
+	client, err := clients.ClientForVersion(&unversioned.GroupVersion{Version: "v1"})
+	if err != nil {
+		return defaultMapper
+	}
+
+	// Check if we have access to server resources
+	apiResources, err := client.Discovery().ServerResources()
+	if err != nil {
+		return defaultMapper
+	}
+
+	availableResources := []unversioned.GroupVersionResource{}
+	for groupVersionString, resourceList := range apiResources {
+		currVersion, err := unversioned.ParseGroupVersion(groupVersionString)
+		if err != nil {
+			return defaultMapper
+		}
+
+		for _, resource := range resourceList.APIResources {
+			availableResources = append(availableResources, currVersion.WithResource(resource.Name))
+		}
+	}
+
+	availableAll := []string{}
+	userResources, _ := defaultMapper.AliasesForResource("all")
+	for _, requestedResource := range userResources {
```

Comment:
This should be a full `GroupResource` (not version) comparison.  Otherwise you could end up enabling resources you didn't mean to from other groups.